File: challenge-4/Permutation.java
/*
* Java Program to find the output of the nPr permutation
* function. nPr = n! / (n - r)! Input Constraints: 
* 2 <= n <= 20
* 0 <= r < n
*/

// Scanner object required for input
import java.util.Scanner;

public class Permutation {

	// Returns true if input constraint is held
	// Returns false if input constraint violated
	private static boolean checkInputConstraint(int n, int r) {

		// Statement on RHS true if n constraint held
		boolean n_constraint = (n >= 2) && (n <= 20);

		// Statement on RHS true if r constraint held
		boolean r_constraint = (r >= 0) && (r < n);

		// If both constraint vars true, then constraint
		// held, otherwise violated
		if (n_constraint && r_constraint) {
			return true;
		} else {
			return false;
		}
	}

	// Returns factorial of given param
	private static long factorialOf(int number) {
		long factorial = 1;

		for (int i = 2; i <= number; i++) {
			factorial *= i;
		}

		return factorial;
	}

	// Driver code
	public static void main(String[] args) {
		
		// Scanner object for input
		Scanner scanner = new Scanner(System.in);

		// Class object to access non-static fields and
		// methods from a static context
		Permutation permutation = new Permutation();

		int n, r;
		System.out.print("Input n: ");
		n = scanner.nextInt();

		System.out.print("Input r: ");
		r = scanner.nextInt();

		// Input constraint failed, exit with code 1
		if (checkInputConstraint(n, r) == false) {
			System.out.println("Error: Input Constraint Violated");
			System.exit(1);
		}

		// Generate output of nPr function.
		double result = (double) factorialOf(n) / factorialOf(n - r);

		System.out.print(n + "P" + r + " = " + result);
	}
}

File: challenge-4/permutation.py
# Python3 Program to find the output of the nPr permutation
# function. nPr = n! / (n - r)! Input Constraints: 
# 2 <= n <= 20
# 0 <= r < n

# Return true if input constraint satisfied
def check_input_constraint(n, r):

	# Statement on RHS becomes true of false. False means
	# constraint violated. True means satisfied.
	n_constraint = (n >= 2) and (n <= 20)

	# Statement of RHS becomes true or false. False if
	# constraint is violated.
	r_constraint = (r >= 0) and (r < n)

	# If both are true, constraint satisfied.
	# else violated.
	if n_constraint and r_constraint:
		return True
	else:
		return False


# Return factorial of param
def factorial_of(num):

	factorial = 1
	for number in range(2, num + 1):
		factorial *= number

	return factorial


# Global, execute first

n = int(input("Input n: "))
r = int(input("Input r: "))

# If input constraint is false, then exit with abnormal
# status code (1)
if check_input_constraint == False:
	print("Error: Input Constraint Violated")
	exit(1)

result = factorial_of(n) / factorial_of(n - r)

print(f'{n}P{r} = {result}')
